README.md
# Divine Equilibrium #

An interdimensional Journey

A Modpack made by `_MasterEnderman_`.

## Story ##

Where am I? I uh... this uh... this can't be good. This can't be good at all. Why is it so dark? I'm not even quite sure which way is up. Okay... I need to think back to this morning to try and figure out how I got here. Let's see now... my alarm woke me up at 7:00 a.m., I got up, took a shower, got dressed, brushed my teeth and went on my daily walk to work. Actually wait, before I left for work I grabbed a potato to eat on my lunch break.

Hmm.. i'm really hungry now, where did my potato go? I don't remember ever eating it, and it doesn't seem to be in my pockets. Where is my potato? =(

## Developer ##

`_MasterEnderman_`

- BuyMeACoffee: <https://www.buymeacoffee.com/masterenderman>
- Curseforge: <https://minecraft.curseforge.com/members/Bitfabrik>
- Discord: <https://discord.gg/JF7x2vG>
- Github: <https://github.com/masterenderman>
- Patreon: <https://www.patreon.com/_masterenderman_>
- Twitch: <https://www.twitch.tv/endermanwhosearchthebeast>
- Twitter: <https://twitter.com/_bitfabrik_>
- Youtube: <https://www.youtube.com/user/xxmasterendermanxx>

## Contributor ##

`al132`

- Akliz Server Referral: <https://www.akliz.net/al132>
- CurseForge: <https://minecraft.curseforge.com/members/al132mc/projects>
- Discord: <https://discord.gg/4swu3fy>
- Github: <https://github.com/al132mc>
- Patreon: <https://www.patreon.com/al132>
- Twitter:<https://twitter.com/al132mc>
- Youtube: <https://www.youtube.com/channel/UCbvuz7eq55a6snmvRI6gXLA>

`Klebestreifen`

- Curseforge: <https://www.curseforge.com/members/klebe99336644>
- Discord: <https://discord.gg/nva5gGt>
- Github: <https://github.com/Klebestreifen>
- Website: <https://klebe.io/>

## Tester ##

- `4oz`
- `Burrito Tiger`
- `Dasky`
- `Dreamin' Boy`
- `Dr. Game`
- `Hurja`
- `JJ`
- `Kr0nos`
- `next_victim`
- `PeachizXD`
- `SahraSingt`
- `Silver_hunt_X`
- `taint19711`
- `Thanos is gay`
- `Vertex101`


CHANGELOG.md
# Changelog #

## 0.2.9 ##

- added easter egg to play pack deathless
- changed gravestone mod
- fixed bone axe recipe
- increased diamond ore generation in the gaia dimension
- made the blood crystal less deadly
- removed hardcore map preset
- removed sulfur generation from gaia dimension

## 0.2.8 ##

- added more QoL recipes
- improved quest description
- fixed scythe not working as hoe in gaia dimension
- fixed progression of the second chapter
- tweaked third chapter

## 0.2.7 ##

- replaced gravestone mod because Corail_31 deleted all his mods
- readded map preset to play the pack on hardcore
- improved topography presets

## 0.2.6 ##

- switched to curseforge version of some mods
- updated README.md

## 0.2.5 ##

- finished chapter 3
- cleaned up modpack for beta release
- fixed everything left in the bugtracker

## 0.2.4 ##

- fixed missing itemstages
- fixed primitive crafting bug

## 0.2.3 ##

- fixed skull recipes
- fixed chest recipe
- fixed reappearing researches
- fixed progression in gaia dimension
- fixed ore gen in gaia dimension
- fixed missing itemstages
- fixed RitualOfTheForest recipe conflicts
- fixed blood infuser crafting time
- improved some recipes

## 0.2.2 ##

- localized all quest titles
- localized all quest descriptions
- improved discord rich presentation
- cleaned up JEI
- worked towards official beta release on curseforge

## 0.2.1 ##

- worked on chapter 3
- finished progression towards the defiled dimension
- tweaked research costs
- improved quest descriptions
- removed hardcore map preset
- cleaned up evil craft
- added blood processing recipes
- added an easter egg to play the pack deathless
- added custom crops
- fixed bone processing
- fixed drying basin
- fixed ancient tree ritual
- tweaked README.md

## 0.2.0 ##

- added defiled dimension
- added hearth containers
- added leaves to oreexcavation whitelist
- added mobspawns to the gaia dimension
- added more quests
- added oregen to the gaia dimension
- added research descriptions
- added sounds to scavenge scripts
- added trophies
- cleaned up JEI
- decreased some quest requirements
- finished gaia progression
- fixed amethyst ore mining level
- fixed most NPEs in elemental research
- fixed some typos in the loading screen messages
- fixed weird numbers in crafting recipes
- fixed world corrupting crash with elemental research
- improved item descriptions
- increased gravel drops from charred cobblestone
- locked default difficulty to hard
- overhauled health system
- replaced old skybox
- replaced starter bone tools
- tweaked the mainmenu
- tweaked the research costs again

## 0.1.3 ##

- removed dimensionaledibles
- added custom teleportation food to travel between Dimensions
- added Roots 3
- removed duplicate content from RootsClassic
- bug fixes (bone gear, sapling ritual)

## 0.1.2 ##

- fixed typo in the quest
- replaced some quests
- fixed a flaw in the progression line
- some proper work on the first dimension
- added forgiving void
- replaced TreeChopper with Ore Excavation
- fixed bone torch consuming when creating charred cobblestone
- added some missing recipes
- added oredict manager script
- added advanced mortars and clockworks phase
- added alternative food for the early game
- tweaked gaia apple height in the gaia dimension

## 0.1.1 ##

- extended Quest Description
- added more basic Quests
- added a few optional quests
- added bone axe to early game
- fixed Quest Descriptions
- updated README.md
- removed Clay Buckets
- added Client Tweaks
- cleaned up JEI
- slightly increased bone meal drop rate from sharp bones
- disabled auto-jump
- tweaked topography template
- added a discord button to the main menu
- added Rich Discord Presence
- added Pack Icon
- updated Elemental Research to fix a lot of bugs
- reduced amount of health a player looses by dying

## 0.1.0 ##

- finished progression towards first dimension
- added more quests (Titles / Descriptions are still W.I.P.)
- tweaked modlist
- optimized loadtime
- refactored scripts
- added proper icons for elemental research
- restructured reseach dependencies
- fixed main menu splash texts

## 0.0.4 ##

- finished skybox in the void
- player can obtain basic wooden materials
- more preparation for the paths
- added way to get hearts back
- refactored scripts
- tweaked the modlist

## 0.0.3 ##

- added custom skybox (WIP)
- added some more mods
- more work towards the first dimension
- started on the "earth path"
- some preparations for the "water path"

## 0.0.2 ##

- Added mods: Nature's Aura, FastLeafDecay
- Updated Mod: Elemental Research (changed air research from 4 jumps to 3)
- Added gitignore entries for most files that change every game launch and aren't required to be sync'd
- Added story to about some of the early quests and tweaked them a bit otherwise (quantities, split one or two up, etc)
- Added some quests after the tree quest with hearth well shards and solidified research via Elemental Research
- Added localizations for contenttweaker blocks, and for player flesh (reported buggy localization to cannibalism dev)

## 0.0.1 ##

- improved some quest descriptions
- skycolor now changes over time
- tweaked spawn chamber color
- tweaked skycolor
- tweaked pebble to cobblestone cost

## 0.0.0 ##

- sorted out the early game
- made a temporary main menu


scripts/Python/CoT_VF_Items.py
import os

"""
This Python programm generates a Script, which generates an Item 
for every entry in a list utilizing ContentTweaker functions.
~ written by _MasterEnderman_
"""

items = {
    # __init__
    "null" : {
        "Pebble" : "pebble",
        "Azurite Pebble" : "azurite",
        "Fuel Pellet" : "pellet",
        "Heavy Ring" : "heavy_ring",
        "Golem Corpse" : "golem_corpse"
    },
    "bark" : {
        "Mysterywood Bark" : "bark_mysterywood",
        "Cherrywood Bark" : "bark_cherrywood",
        "Ancient Bark" : "bark_ancient"
    },
    "icon" : {
        "Primitive Crafting Icon" : "primcraft"
    },
    "materials" : {
        "Terra Ingot" : "terra_ingot"
    },
    "solidified_knowledge" : {
        "Solidified Fire Knowledge" : "solidified_fire_knowledge",
        "Solidified Water Knowledge" : "solidified_water_knowledge",
        "Solidified Air Knowledge" : "solidified_air_knowledge",
        "Solidified Earth Knowledge" : "solidified_earth_knowledge"
    },
    "tools" : {
        "The Hwaxun" : "hwaxun",
        "Stone Pickaxe Head" : "head_stone_pickaxe",
        "Stone Axe Head" : "head_stone_axe",
        "Stone Shovel Head" : "head_stone_shovel",
        "Stone Sword Blade" : "head_stone_sword"
    }
}

head = """#priority 1000
#loader contenttweaker

import mods.contenttweaker.Item;
import mods.contenttweaker.LocalizedNameSupplier;
import mods.contenttweaker.ResourceLocation;
import mods.contenttweaker.VanillaFactory;

print("Loading VF_Items");\n
"""

with open(os.getcwd()+r"\.minecraft\scripts\ContentTweaker\VanillaFactory\VF_Items.zs", "r+") as f:
    f.truncate(0)
    f.write(head)
    for key, value in items.items():
        f.write("//"+key+"\n")
        for lang, name in value.items():
            f.write("var "+name+' as Item = VanillaFactory.createItem("'+name+'");\n')
            f.write(name+'.setLocalizedNameSupplier(function(itemStack) {return "'+lang+'";});\n')
            if (key != "null"):
                f.write(name+'.textureLocation = ResourceLocation.create("contenttweaker:items/'+key+"/"+name+'");\n')
            f.write(name+".register();\n\n")


scripts/Python/intConverter.py
def rgb2hex(r, g, b):
    return "#{:02x}{:02x}{:02x}".format(r, g, b)
    
intintintlist = {
    "hydrogen" : [0, 0, 255],
    "helium" : [255, 0, 0],
    "lithium" : [40, 158, 86],
    "beryllium" : [184, 199, 224],
    "boron" : [154, 176, 226],
    "carbon" : [59, 60, 63],
    "nitrogen" : [66, 123, 214],
    "oxygen" : [229, 220, 156],
    "fluorine" : [204, 186, 55],
    "neon" : [87, 229, 16],
    "sodium" : [211, 198, 131],
    "magnesium" : [237, 178, 173],
    "aluminum" : [247, 110, 69],
    "silicon" : [173, 178, 121],
    "phosphorus" : [234, 98, 132],
    "sulfur" : [145, 158, 6],
    "chlorine" : [77, 102, 28],
    "argon" : [119, 117, 255],
    "potassium" : [198, 152, 95],
    "calcium" : [219, 210, 199],
    "scandium" : [252, 255, 99],
    "titanium" : [99, 255, 115],
    "vanadium" : [195, 186, 242],
    "chromium" : [236, 237, 218],
    "manganese" : [225, 186, 242],
    "iron" : [128, 128, 128],
    "cobalt" : [17, 114, 198],
    "nickel" : [198, 157, 162],
    "copper" : [255, 154, 30],
    "zinc" : [189, 196, 141],
    "gallium" : [122, 20, 49],
    "germanium" : [104, 172, 255],
    "arsenic" : [62, 145, 76],
    "selenium" : [116, 62, 145],
    "bromine" : [77, 160, 0],
    "krypton" : [229, 151, 50],
    "rubidium" : [15, 61, 40],
    "strontium" : [206, 88, 24],
    "yttrium" : [206, 179, 24],
    "zirconium" : [127, 80, 22],
    "niobium" : [2, 29, 255],
    "molybdenum" : [39, 0, 48],
    "technetium" : [72, 170, 63],
    "ruthenium" : [255, 240, 86],
    "rhodium" : [255, 0, 80],
    "palladium" : [0, 255, 169],
    "silver" : [226, 217, 206],
    "cadmium" : [160, 147, 115],
    "indium" : [163, 230, 255],
    "tin" : [132, 161, 206],
    "antimony" : [193, 40, 58],
    "tellurium" : [39, 91, 26],
    "iodine" : [62, 17, 63],
    "xenon" : [196, 51, 204],
    "cesium" : [255, 148, 0],
    "barium" : [0, 219, 179],
    "lanthanum" : [188, 253, 255],
    "cerium" : [255, 254, 211],
    "praseodymium" : [255, 161, 0],
    "neodymium" : [38, 28, 11],
    "promethium" : [105, 175, 123],
    "samarium" : [73, 69, 73],
    "europium" : [27, 211, 45],
    "gadolinium" : [123, 50, 208],
    "terbium" : [3, 37, 118],
    "dysprosium" : [73, 0, 219],
    "holmium" : [62, 255, 56],
    "erbium" : [194, 214, 215],
    "thulium" : [234, 178, 178],
    "ytterbium" : [255, 76, 219],
    "lutetium" : [175, 0, 219],
    "hafnium" : [69, 81, 233],
    "tantalum" : [108, 142, 110],
    "tungsten" : [120, 128, 140],
    "rhenium" : [199, 226, 89],
    "osmium" : [102, 129, 173],
    "iridium" : [215, 242, 238],
    "platinum" : [114, 202, 229],
    "gold" : [255, 255, 0],
    "mercury" : [160, 159, 157],
    "thallium" : [103, 50, 25],
    "lead" : [186, 135, 193],
    "bismuth" : [252, 171, 40],
    "polonium" : [138, 87, 85],
    "astatine" : [120, 128, 213],
    "radon" : [76, 66, 179],
    "francium" : [81, 114, 198],
    "radium" : [255, 181, 221],
    "actinium" : [14, 182, 145],
    "thorium" : [56, 79, 75],
    "protactinium" : [204, 233, 2],
    "uranium" : [93, 178, 19],
    "neptunium" : [32, 20, 158],
    "plutonium" : [211, 211, 209],
    "americium" : [237, 124, 75],
    "curium" : [229, 110, 149],
    "berkelium" : [44, 66, 49],
    "californium" : [175, 182, 16],
    "einsteinium" : [192, 210, 95],
    "fermium" : [74, 226, 83],
    "mendelevium" : [175, 176, 249],
    "nobelium" : [94, 44, 52],
    "lawrencium" : [216, 45, 92],
    "rutherfordium" : [240, 61, 22],
    "dubnium" : [11, 112, 108],
    "seaborgium" : [158, 49, 74],
    "bohrium" : [166, 251, 51],
    "hassium" : [78, 5, 51],
    "meitnerium" : [169, 138, 37],
    "darmstadtium" : [14, 144, 190],
    "roentgenium" : [150, 90, 90],
    "copernicium" : [160, 40, 240],
    "nihonium" : [220, 250, 180],
    "flerovium" : [200, 180, 254],
    "moscovium" : [250, 180, 200],
    "livermorium" : [250, 250, 200],
    "tennessine" : [150, 250, 250],
    "oganesson" : [250, 150, 250]
}

index = 1

for key, value in intintintlist.items():
    print('"'+key+"_"+str(index)+'" : ', 'r"'+rgb2hex(value[0],value[1],value[2])+'",')
    index +=1

hexlist = {
    #"hydrogen_1":  r"#0000ff",
    #"helium_2":  r"#ff0000",
    "lithium_3":  r"#289e56",
    "beryllium_4":  r"#b8c7e0",
    "boron_5":  r"#9ab0e2",
    #"carbon_6":  r"#3b3c3f",
    #"nitrogen_7":  r"#427bd6",
    #"oxygen_8":  r"#e5dc9c",
    #"fluorine_9":  r"#ccba37",
    #"neon_10":  r"#57e510",
    "sodium_11":  r"#d3c683",
    "magnesium_12":  r"#edb2ad",
    "aluminum_13":  r"#f76e45",
    "silicon_14":  r"#adb279",
    #"phosphorus_15":  r"#ea6284",
    #"sulfur_16":  r"#919e06",
    #"chlorine_17":  r"#4d661c",
    #"argon_18":  r"#7775ff",
    "potassium_19":  r"#c6985f",
    "calcium_20":  r"#dbd2c7",
    "scandium_21":  r"#fcff63",
    "titanium_22":  r"#63ff73",
    "vanadium_23":  r"#c3baf2",
    "chromium_24":  r"#ecedda",
    "manganese_25":  r"#e1baf2",
    "iron_26":  r"#808080",
    "cobalt_27":  r"#1172c6",
    "nickel_28":  r"#c69da2",
    "copper_29":  r"#ff9a1e",
    "zinc_30":  r"#bdc48d",
    "gallium_31":  r"#7a1431",
    "germanium_32":  r"#68acff",
    "arsenic_33":  r"#3e914c",
    "selenium_34":  r"#743e91",
    #"bromine_35":  r"#4da000",
    #"krypton_36":  r"#e59732",
    "rubidium_37":  r"#0f3d28",
    "strontium_38":  r"#ce5818",
    "yttrium_39":  r"#ceb318",
    "zirconium_40":  r"#7f5016",
    "niobium_41":  r"#021dff",
    "molybdenum_42":  r"#270030",
    "technetium_43":  r"#48aa3f",
    "ruthenium_44":  r"#fff056",
    "rhodium_45":  r"#ff0050",
    "palladium_46":  r"#00ffa9",
    "silver_47":  r"#e2d9ce",
    "cadmium_48":  r"#a09373",
    "indium_49":  r"#a3e6ff",
    "tin_50":  r"#84a1ce",
    "antimony_51":  r"#c1283a",
    "tellurium_52":  r"#275b1a",
    #"iodine_53":  r"#3e113f",
    #"xenon_54":  r"#c433cc",
    "cesium_55":  r"#ff9400",
    "barium_56":  r"#00dbb3",
    "lanthanum_57":  r"#bcfdff",
    "cerium_58":  r"#fffed3",
    "praseodymium_59":  r"#ffa100",
    "neodymium_60":  r"#261c0b",
    "promethium_61":  r"#69af7b",
    "samarium_62":  r"#494549",
    "europium_63":  r"#1bd32d",
    "gadolinium_64":  r"#7b32d0",
    "terbium_65":  r"#032576",
    "dysprosium_66":  r"#4900db",
    "holmium_67":  r"#3eff38",
    "erbium_68":  r"#c2d6d7",
    "thulium_69":  r"#eab2b2",
    "ytterbium_70":  r"#ff4cdb",
    "lutetium_71":  r"#af00db",
    "hafnium_72":  r"#4551e9",
    "tantalum_73":  r"#6c8e6e",
    "tungsten_74":  r"#78808c",
    "rhenium_75":  r"#c7e259",
    "osmium_76":  r"#6681ad",
    "iridium_77":  r"#d7f2ee",
    "platinum_78":  r"#72cae5",
    "gold_79":  r"#ffff00",
    "mercury_80":  r"#a09f9d",
    "thallium_81":  r"#673219",
    "lead_82":  r"#ba87c1",
    "bismuth_83":  r"#fcab28",
    "polonium_84":  r"#8a5755",
    "astatine_85":  r"#7880d5",
    #"radon_86":  r"#4c42b3",
    "francium_87":  r"#5172c6",
    "radium_88":  r"#ffb5dd",
    "actinium_89":  r"#0eb691",
    "thorium_90":  r"#384f4b",
    "protactinium_91":  r"#cce902",
    "uranium_92":  r"#5db213",
    "neptunium_93":  r"#20149e",
    "plutonium_94":  r"#d3d3d1",
    "americium_95":  r"#ed7c4b",
    "curium_96":  r"#e56e95",
    "berkelium_97":  r"#2c4231",
    "californium_98":  r"#afb610",
    "einsteinium_99":  r"#c0d25f",
    "fermium_100":  r"#4ae253",
    "mendelevium_101":  r"#afb0f9",
    "nobelium_102":  r"#5e2c34",
    "lawrencium_103":  r"#d82d5c",
    "rutherfordium_104":  r"#f03d16",
    "dubnium_105":  r"#0b706c",
    "seaborgium_106":  r"#9e314a",
    "bohrium_107":  r"#a6fb33",
    "hassium_108":  r"#4e0533",
    "meitnerium_109":  r"#a98a25",
    "darmstadtium_110":  r"#0e90be",
    "roentgenium_111":  r"#965a5a",
    "copernicium_112":  r"#a028f0",
    "nihonium_113":  r"#dcfab4",
    "flerovium_114":  r"#c8b4fe",
    "moscovium_115":  r"#fab4c8",
    "livermorium_116":  r"#fafac8",
    "tennessine_117":  r"#96fafa",
    "oganesson_118":  r"#fa96fa"
}


scripts/Python/CoT_VF_Blocks.py
import os

"""
This Python programm generates a Script, which generates a Block 
for every entry in a list utilizing ContentTweaker functions.
~ written by _MasterEnderman_
"""

#blockmaterial = [Air,Grass,Ground,Wood,Rock,Iron,Anvil,Water,Lava,Leaves,Plants,Vine,Sponge,Cloth,Fire,Sand,Circuits,Carpet,Glass,Redstone_Light,TNT,Coral,Ice,Packed_Ice,Crafted_Snow,Cactus,Clay,Gourd,Dragon_Egg,Portal,Cake,Web]
#soundtypes = [Wood,Ground,Plant,Stone,Metal,Glass,Cloth,Sand,Snow,Ladder,Anvil,Slime?]

blocks = {
    # ID // Material // Hardness // Resistence // Tool // Tool Lvl // Sound
    "null" : [
        ["spawn_chamber", "rock", 10, 30, "pickaxe", 0, "stone"],
        ["charred_cobblestone", "rock", 2, 30, "pickaxe", 0, "stone"],
        ["flesh_block", "rock", 2, 10, "pickaxe", 0, "ground"]
    ]
}

head = """#priority 1000
#loader contenttweaker

import mods.contenttweaker.Block;
import mods.contenttweaker.ResourceLocation;
import mods.contenttweaker.VanillaFactory;

print("Loading VF_Blocks");\n
"""

with open(os.getcwd()+r"\.minecraft\scripts\ContentTweaker\VanillaFactory\VF_Blocks.zs", "r+") as f:
    f.truncate(0)
    f.write(head)
    for key, value in blocks.items():
        for i in value:
            f.write("var "+i[0]+' = VanillaFactory.createBlock("'+i[0]+'", <blockmaterial:'+i[1]+'>);\n')
            if (key != "null"):
                f.write(i[0]+'.textureLocation = ResourceLocation.create("contenttweaker:blocks/'+key+"/"+i[0]+'");\n')
            f.write(i[0]+'.setBlockHardness('+str(float(i[2]))+');\n')
            f.write(i[0]+'.setBlockResistance('+str(float(i[3]))+');\n')
            f.write(i[0]+'.setToolClass("'+i[4]+'");\n')
            f.write(i[0]+'.setToolLevel('+str(int(i[5]))+');\n')
            f.write(i[0]+'.setBlockSoundType(<soundtype:'+i[6]+'>);\n')
            f.write(i[0]+'.register();\n\n')


scripts/Python/CoT_VF_Fluids.py
import os

"""
This Python programm generates a Script, which generates a Fluid 
for every entry in a list utilizing ContentTweaker funtions.
~ written by _MasterEnderman_
"""

fluids = [
    # Liquid Name // Color (hex) // Density (Walkspeed) [Default:1000] // Luminosity (Lightlevel) (int) [Default: 0] // Temperature (int) [Default: 300]
    ["frankencense_essence", "#C26D23", 1000, 1, 290],
    ["liquid_darkness", "#111111", 300, 0, 10]
]

head = """#priority 1000
#loader contenttweaker

import mods.contenttweaker.VanillaFactory;
import mods.contenttweaker.Fluid;

print("Loading VF_Fluids");\n
"""

with open(os.getcwd()+r"\.minecraft\scripts\ContentTweaker\VanillaFactory\VF_Fluids.zs", "r+") as f:
    f.truncate(0)
    f.write(head)
    for i in range(len(fluids)):
        f.write("var "+fluids[i][0]+' as Fluid = VanillaFactory.createFluid("'+fluids[i][0]+'",'+str(int(fluids[i][1].replace(r"#",""),16))+');\n')
        f.write(fluids[i][0]+".setDensity("+str(fluids[i][2])+");\n")
        f.write(fluids[i][0]+".setLuminosity("+str(fluids[i][3])+");\n")
        f.write(fluids[i][0]+".setTemperature("+str(fluids[i][4])+");\n")
        f.write(fluids[i][0]+".register();\n\n")


scripts/Python/CoT_VF_FoodItems.py
import os

"""
This Python programm generates a Script, which generates a Food Item 
for every entry in a list utilizing ContentTweaker functions.
~ written by _MasterEnderman_
"""

foods = {
    # Name // healAmount // saturation [Default: 0.6]
    "apple" : {
        "mystical" : {
            "hunger" : 4,
            "saturation" : 0.9
        }
    }
}

head = """#priority 1000
#loader contenttweaker

import mods.contenttweaker.ItemFood;
import mods.contenttweaker.ResourceLocation;
import mods.contenttweaker.VanillaFactory;

print("Loading VF_FoodItems");\n
"""

with open(os.getcwd()+r"\.minecraft\scripts\ContentTweaker\VanillaFactory\VF_FoodItems.zs", "r+") as f:
    f.truncate(0)
    f.write(head)
    for category, types in foods.items():
        f.write("//"+category+"\n")
        for food, data in types.items():
            itemname = category+"_"+food
            f.write("var "+itemname+' = VanillaFactory.createItemFood("'+itemname+'",'+str(data["hunger"])+');\n')
            f.write(itemname+".setSaturation("+str(data["saturation"])+");\n")
            f.write(itemname+'.textureLocation = ResourceLocation.create("contenttweaker:items/food/'+itemname+'");\n')
            f.write(itemname+".register();\n\n")


scripts/Python/CoT_MS.py
import os

"""
This Python programm generates a ZenScript, which automates the Creation of the required
files utilizing ContentTweaker's MaterialSystem.
~ written by _MasterEnderman_
"""

parts = [ #Custom Parts you want to add have to be specified here
    #Part name // Ore Dict name // TAG:type
    #["dust_tiny", "dustTiny", "crafting"]
]

materials = [ #Custom Materials you want to add have to be specified here
    #Material name // Localized Name // Color (hex) // Gem? // [Crafting(Ingot,Plate),Processing(Dust,Ore),Fluid]
    ["brass", "Brass", "#D2983A", False, [False, False, True]],
    ["thaumium", "Thaumium", "#4A3D71", False, [False, False, True]]
]

crafting = [ #Default ContentTweaker values, disable a part by commenting it out
    #Part name // Ore Dict name
    #["block","block"],
	["nugget","nugget"],
	["ingot","ingot"],
	["plate","plate"],
    #["dense_plate","plateDense"],
	#["gear","gear"],
    #["beam","beam"],
    #["bolt","bolt"],
    #["casing","casing"],
    #["rod","rod"],
	#["dust","dust"]
]

processing = [ #Default ContentTweaker values, disable a part by commenting it out
    #Part name // Ore Dict name
    #["cluster","cluster"],
   	#["clump","clump"],
   	#["crushed_ore","oreCrushed"],
   	#["dirty_dust","dustDirty"],
    #["crystal","crystal"],
    #["shard","shard"],
    #["ore","ore"]
]

gem = [  #Default ContentTweaker values, disable a part by commenting it out
    #Part name // Ore Dict name
    ["ore", "ore"]
]

for i in range(len(parts)):
    temp_list = [parts[i][0], parts[i][1]]
    if parts[i][2] == "CRAFTING":
        crafting.append(temp_list)
    elif parts[i][2] == "PROCESSING":
        processing.append(temp_list)
    elif parts[i][2] == "GEM":
    	gem.append(temp_list)

MS_Definition = """#priority 100000
#loader contenttweaker

import mods.contenttweaker.Color;
import mods.contenttweaker.Material;
import mods.contenttweaker.MaterialPartData;
import mods.contenttweaker.MaterialSystem;

print("Loading MS_Definition");\n
"""

funtions = """function addDefaultMoltenData(moltenData as MaterialPartData) {
	moltenData.addDataValue("density", "4000");
	moltenData.addDataValue("viscosity", "3000");
	moltenData.addDataValue("temperature", "1500");
	moltenData.addDataValue("vaporize", "false");
}

function addDefaultOreData(oreData as MaterialPartData) {
	oreData.addDataValue("variants", "minecraft:stone");
	oreData.addDataValue("hardness", "5");
	oreData.addDataValue("resistance", "15");
	oreData.addDataValue("harvestTool", "pickaxe");
	oreData.addDataValue("harvestLevel", "1");
}
"""

space = "    "

with open(os.getcwd()+r"\.minecraft\scripts\ContentTweaker\MaterialSystem\MS_Definition.zs", "r+") as f:
    f.truncate(0)
    f.write(MS_Definition)
    for i in range(len(parts)):
        f.write("var "+parts[i][0]+' = mods.contenttweaker.MaterialSystem.getPartBuilder().setName("'+parts[i][0]+
            '").setPartType(mods.contenttweaker.MaterialSystem.getPartType("item")).setOreDictName("'+parts[i][1]+
            '").build();\n')
    if len(materials) > 0:
        f.write("\nstatic materials as Material[string] = {\n")
        for i in range(len(materials)):
            if i == len(materials)-1:
                f.write(space+'"'+materials[i][0]+'": MaterialSystem.getMaterialBuilder().setName("'+materials[i][1]+
                    '").setColor('+str(int(materials[i][2].replace(r"#",""),16))+').build()\n')
            else:
                f.write(space+'"'+materials[i][0]+'": MaterialSystem.getMaterialBuilder().setName("'+materials[i][1]+
                    '").setColor('+str(int(materials[i][2].replace(r"#",""),16))+').build(),\n')
        f.write("};\n")
    f.write("\n"+funtions)

MS_Registration = """#priority 99999
#loader contenttweaker

import mods.contenttweaker.Material;
import mods.contenttweaker.MaterialPartData;
import mods.contenttweaker.MaterialSystem;

import scripts.ContentTweaker.MaterialSystem.MS_Definition.addDefaultMoltenData;
import scripts.ContentTweaker.MaterialSystem.MS_Definition.addDefaultOreData;
import scripts.ContentTweaker.MaterialSystem.MS_Definition.materials;

print("Loading MS_Registration");
"""

with open(os.getcwd()+r"\.minecraft\scripts\ContentTweaker\MaterialSystem\MS_Registration.zs", "r+") as f:
    f.truncate(0)
    f.write(MS_Registration)
    for i in range(len(materials)):
        f.write("\n//"+materials[i][0]+"\n\n")
        if not materials[i][3]:
        	registerMetal = []
	        if materials[i][4][0]:
	            for j in range(len(crafting)):
	                registerMetal.append(crafting[j][0])
	        if materials[i][4][1]:
	            for k in range(len(processing)):
	                if processing[k][0] == "ore":
	                    f.write("var "+materials[i][0]+"OreData as MaterialPartData = " +
	                            "materials."+materials[i][0]+'.registerPart("ore").getData();\n')
	                    f.write("addDefaultOreData("+materials[i][0]+"OreData);\n\n")
	                else:
	                    registerMetal.append(processing[k][0])
	        if materials[i][4][2]:
	            f.write("var "+materials[i][0]+"FluidData as MaterialPartData = " +
	                    "materials."+materials[i][0]+'.registerPart("molten").getData();\n')
	            f.write("addDefaultMoltenData("+materials[i][0]+"FluidData);\n\n")
	        if materials[i][4][0] or materials[i][4][1]:
	            f.write('materials.'+materials[i][0]+'.registerParts(' +
	                    str(registerMetal).replace("'", '"')+' as string[]);\n')
        else:
            registerGem = []
            for k in range(len(gem)):
	                if gem[k][0] == "ore":
	                    f.write("var "+materials[i][0]+"OreData as MaterialPartData = " +
	                            "materials."+materials[i][0]+'.registerPart("ore").getData();\n')
	                    f.write(
	                        "addDefaultOreData("+materials[i][0]+"OreData);\n\n")
	                else:
	                    registerGem.append(gem[k][0])
            f.write('materials.'+materials[i][0]+'.registerParts('+str(registerGem).replace("'",'"')+' as string[]);\n')


metalObjects = """#priority 10000

import crafttweaker.item.IItemStack;
import crafttweaker.item.IIngredient;
import crafttweaker.liquid.ILiquidDefinition;
import crafttweaker.liquid.ILiquidStack;
import crafttweaker.oredict.IOreDict;
import crafttweaker.oredict.IOreDictEntry;

print("Loading metalObjects");

global metalObjects as IOreDictEntry[string][string] = {
"""

with open(os.getcwd()+r"\.minecraft\scripts\Crafttweaker\metalObjects.zs", "r+") as f:
    f.truncate(0)
    f.write(metalObjects)
    for i in range(len(materials)):
        if not materials[i][3]:
            f.write(space+'"'+materials[i][0]+'" : {\n')
            if materials[i][4][0]:
                for n in range(len(crafting)):
                    if n == len(crafting)-1 and not materials[i][4][1]:
                        f.write(space*2+'"'+crafting[n][0]+'" : <ore:' +
                            crafting[n][1]+materials[i][1].replace(" ","")+">\n")
                    else:
                        f.write(space*2+'"'+crafting[n][0]+'" : <ore:' +
                                crafting[n][1]+materials[i][1].replace(" ", "")+">,\n")
            if materials[i][4][1]:
                for n in range(len(processing)):
                    if n == len(processing)-1:
                        f.write(space*2+'"'+processing[n][0]+'" : <ore:' +
                                processing[n][1]+materials[i][1].replace(" ", "")+">\n")
                    else:
                        f.write(space*2+'"'+processing[n][0]+'" : <ore:' +
                                processing[n][1]+materials[i][1].replace(" ", "")+">,\n")
            if i < len(materials)-1:
                if materials[i+1][3]:
                    f.write(space+"}\n")
                else:
                    f.write(space+"},\n")
            else:
                f.write(space+"}\n")
    f.write("};")
